### ISSUE TITLE
Hyprland calls linux_dmabuf_feedback_handle_main_device twice

### ISSUE THREAD

**Issue by ObsidianChickenz** [hyprwm/Hyprland#294]
This sounds like a bug in xdg-desktop-portal-wlr, but this doesn't occur in other compositors, only Hyprland. I'm running NixOS unstable, if that helps at all.

Steps to reproduce:
- Install Hyprland by adding it to the list of installed packages
- Enable xdg-desktop-portal-wlr and pipewire by setting the required options in configuration.nix
- Start xdg-desktop-portal-wlr by running `systemctl --user start xdg-desktop-portal-wlr`

Expected outcome: xdg-desktop-portal-wlr starts as usual.
Noted outcome: xdg-desktop-portal-wlr dumps core and crashes

I'll link to the issue in xdg-desktop-portal-wlr here: [https://github.com/emersion/xdg-desktop-portal-wlr/issues/216](https://github.com/emersion/xdg-desktop-portal-wlr/issues/216)

Apparently Hyprland sends feedback twice, which is unusual.

I'm not entirely sure what's going on, just trying to help where possible. If anyone needs me to post any logs, I can do so, just let me know.

**Comment by fufexan** [comment]
@ObsidianChickenz is the xdpw service running? Or does it fail on startup? Mine was failing with [this](https://github.com/hyprwm/Hyprland/files/9031255/message.txt) and I had to apply a patch to make it work again.

**Comment by ObsidianChickenz** [comment]
When I run `systemctl --user status xdg-desktop-portal-wlr.service`, it outputs:
`○ xdg-desktop-portal-wlr.service - Portal service (wlroots implementation)
     Loaded: loaded (/etc/systemd/user/xdg-desktop-portal-wlr.service; static)
    Drop-In: /nix/store/m5nfp51h4sbdilfyvfd3k25mgkqrzgss-user-units/xdg-desktop-portal-wlr.service.d
             └─overrides.conf
     Active: inactive (dead)`

**Comment by fufexan** [comment]
@ObsidianChickenz try using this patch, worked for me.
```diff
diff --git a/src/screencast/wlr_screencast.c b/src/screencast/wlr_screencast.c
index 8d64553..ebfb5e2 100644
--- a/src/screencast/wlr_screencast.c
+++ b/src/screencast/wlr_screencast.c
@@ -609,7 +609,7 @@ static void linux_dmabuf_feedback_handle_main_device(void *data,
 
 	logprint(DEBUG, "wlroots: linux_dmabuf_feedback_handle_main_device called");
 
-	assert(ctx->gbm == NULL);
+	if (ctx->gbm != NULL) return;
 
 	dev_t device;
 	assert(device_arr->size == sizeof(device));
```
Let me know if you need help with it.

Also, the issue is still there. I try to screenshare but it doesn't pop up `slurp` so that I can select my output.

**Comment by ObsidianChickenz** [comment]
How would I go about this on NixOS?

**Comment by fufexan** [comment]
```nix
# inside your config
nixpkgs.overlays = [
  (final: prev: {
    xdg-desktop-portal-wlr = prev.xdg-desktop-portal-wlr.overrideAttrs (_: {
      patches = [./xdpw-crash.patch];
    });
  })
];
```
And have the patch with the name `xdpw-crash.patch` in the same directory. Then rebuild and `systemctl --user restart xdg-desktop-portal-wlr` and it should be up.

**Comment by ObsidianChickenz** [comment]
This worked, and it started now, but I still wonder why patching is required in the first place. Is there any way to make it so that Hyprland doesn't conflict with xdpw?

**Comment by fufexan** [comment]
> I still wonder why patching is required in the first place

Me too, idk.

> Is there any way to make it so that Hyprland doesn't conflict with xdpw?

Again, idk. It used to work a few days ago.

**Comment by ObsidianChickenz** [comment]
It seems that Hyprland calls `linux_dmabuf_feedback_handle_main_device` twice, which creates the issue. I don't know where in the code it does this, but that's what I'm getting from the logs in the other issue I sent.

**Comment by vaxerski** [comment]
it indeed is that. Why? no clue. Could I probably find why? Most likely. Will I do it? Since it's not a really important issue (one patch fixes the issue), probably not loterally right now, but I will look into it one day.

**Comment by ObsidianChickenz** [comment]
I'll reopen this until it gets patched, and I'll rename the issue.

**Comment by ObsidianChickenz** [comment]
Is there anywhere in particular where these things are called? I could look so you don't have to, but this is also my first time contributing.

**Comment by fufexan** [comment]
@ObsidianChickenz you could clone the repo and grep that function with `rg` or something.

**Comment by vaxerski** [comment]
well you can debug hyprland according to the instructions in the wiki and put a breakpoint on the func no?

**Comment by fufexan** [comment]
I've actually grepped that function and it's nowhere to be found lol.

**Comment by vaxerski** [comment]
it's in wlroots.

**Comment by ObsidianChickenz** [comment]
`rg` returned nothing when searching for `linux_dmabuf_feedback_handle_main_device`. I searched the whole repo. Would wlroots be outside the repo?

**Comment by vaxerski** [comment]
wlroots is in subprojects/

**Comment by ObsidianChickenz** [comment]
When I cloned the repo, the wlroots directory was empty. Did I mess up somewhere?

**Comment by fufexan** [comment]
Should've cloned with `--recursive` to get it first time. But you can get it with `git submodule init subprojects/wlroots`

**Comment by ObsidianChickenz** [comment]
Recloned with recursive, found a function called `linux_dmabuf_feedback_v1_handle_main_device`, which is very similar, in `subprojects/wlroots/backend/wayland/backend.c`. However, I can't find any duplicates anywhere.

**Comment by vaxerski** [comment]
see what function calls that and what function calls that again and again until u get to hyprland and maybe you'll find the culprit.

**Comment by ObsidianChickenz** [comment]
I found out that `linux_dmabuf_feedback_handle_main_device` is actually a function in xdpw, not wlroots. I'm trying to find what calls the function there now.

**Comment by ObsidianChickenz** [comment]
No amound of grepping helps, I can't find where `linux_dmabuf_feedback_handle_main_device` is called, whether in xdpw, wlroots, or Hyprland.

**Comment by ObsidianChickenz** [comment]
How would I debug Hyprland on NixOS? I think I might be starting to get somewhere.

**Comment by fufexan** [comment]
@ObsidianChickenz you can install the `hyprland-debug` package.

**Comment by ObsidianChickenz** [comment]
Would this be in the normal Nixpkgs repo or in the flake?

**Comment by viperML** [comment]
The flake: https://github.com/hyprwm/Hyprland/blob/b7b3344188e3ebd2baed14bc288ee42a1ae8d6f2/flake.nix#L38

**Comment by ObsidianChickenz** [comment]
The debug log doesn't update when xdpw is started. Is there any other way that I may be able to see how Hyprland and xdpw interact?

**Comment by vaxerski** [comment]
they dont. Hyprland exposes a protocol and xdpw uses it. The protocol itself is handled by wlroots.

**Comment by ObsidianChickenz** [comment]
Is there any log of when Hyprland exposes the protocol?

**Comment by vaxerski** [comment]
no. It's a wlroots manager afaik, see `CCompositor::CCompositor()`

**Comment by ObsidianChickenz** [comment]
I've traced `linux_dmabuf_feedback_v1_handle_main_device` all the way back to Compositor.cpp without finding any duplicates. Can someone else verify this?

**Comment by ObsidianChickenz** [comment]
Is `linux_dmabuf_feedback_v1_handle_main_device` what sends the feedback, for sure?

**Comment by vaxerski** [comment]
I don't know anything more about the subject. I'd guess so.

**Comment by vaxerski** [comment]
fixed in 217ff1f4011336e51bc65bd98fd7b931547ac45c
